microcoap/main.c

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright (C) 2015 HAW Hamburg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @{
+ *
+ * @file
+ * @brief       microcoap example server
+ *
+ * @author      Lotte Steenbrink <lotte.steenbrink@haw-hamburg.de>
+ *
+ * @}
+ */
+
+#include <stdio.h>
+
+#include "udp.h"
+#include "net_help.h"
+#include "net_if.h"
+#include "board_uart0.h"
+#include "shell.h"
+#include "shell_commands.h"
+#include "thread.h"
+#include "posix_io.h"
+
+#include <coap.h>
+
+#define ENABLE_DEBUG    (1)
+#include "debug.h"
+
+#define PORT 5683
+#define BUFSZ 128
+
+#define RCV_MSG_Q_SIZE      (64)
+
+static void *_microcoap_server_thread(void *arg);
+
+msg_t msg_q[RCV_MSG_Q_SIZE];
+char _rcv_stack_buf[KERNEL_CONF_STACKSIZE_MAIN];
+
+static ipv6_addr_t prefix;
+int sock_rcv, if_id;
+sockaddr6_t sa_rcv;
+uint8_t buf[BUFSZ];
+uint8_t scratch_raw[BUFSZ];
+coap_rw_buffer_t scratch_buf = {scratch_raw, sizeof(scratch_raw)};
+
+static void _init_tlayer(void);
+static uint16_t get_hw_addr(void);
+
+int main(void)
+{
+
+    DEBUG("Starting example microcoap server...\n");
+
+    _init_tlayer();
+    thread_create(_rcv_stack_buf, KERNEL_CONF_STACKSIZE_MAIN, PRIORITY_MAIN, CREATE_STACKTEST, _microcoap_server_thread, NULL ,"_microcoap_server_thread");
+
+    /* Open the UART0 for the shell */
+    posix_open(uart0_handler_pid, 0);
+
+    printf("\n\t\t\tWelcome to RIOT\n\n");
+
+    shell_t shell;
+    shell_init(&shell, NULL, UART0_BUFSIZE, uart0_readc, uart0_putc);
```

Comment:
With the CPU ID it is randomized, but for each instance it will stay the same: say you have two samr21-xpro boards, than for board 1 the address will always be e.g. `0x4a2c` and for board 2 it will always be `0x5fb4`. Only for native where you effectively "destroy" your board everytime you kill the process it is running in the ID changes. But you can circumvent this with the `-i` parameter to your TERMFLAGS. E. g.:

```
$ ./examples/default/bin/native/default.elf tap0
[...]
CPU ID: 000000 82 33 00 00
$ ./examples/default/bin/native/default.elf tap0
[...]
CPU ID: 000000 86 33 00 00
$ ./examples/default/bin/native/default.elf tap0 -i 1234
[...]
CPU ID: 000000 d2 04 00 00
$ ./examples/default/bin/native/default.elf tap0 -i 1234
[...]
CPU ID: 000000 d2 04 00 00
```

Problem is: for a lot of the older boards (e.g. the msba2) this feature is not implemented yet.